### PR TITLE
ROX-24725: Sensor limit simultaneous scans per image

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Scan events will miss the Sensor scan cache when an image already exists in the cache and then events for the same image digest but different names are received in quick succession.  A scan would be requested for each new event.

This occurs because the ‘signal’ that is used in Sensor to control access to the cached image only applies to the first image name observed, concurrency mechanisms are not used on subsequent events that reference different names.

This PR:
- Ensures at any given time only a single scan per image (per sensor) will be active.
  - By acquiring the `cacheValue` lock before the scan request is sent
- Ensures that the same image isn't sent for enrichment multiple times when new names are observed.
  - By always appending to the cached image names regardless of what Central returns.

This addresses a cache miss scenario that was amplified by the Unqualified Search Registries feature (which changes image names to match what the runtime provided)

The enhancement will only take effect if both `ROX_UNQUALIFIED_SEARCH_REGISTRIES` and `ROX_SENSOR_SINGLE_SCAN` are set to true. ROX-24641 has been created to remove the dependency on `ROX_UNQUALIFIED_SEARCH_REGISTRIES` in the future if the feature works well in the field.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [x] Determined and documented upgrade steps
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
Unit tests and on an infra OCP cluster, installed ACS, scaled scanner to zero replicas to 'simulate' failures, and created various deployments for the same image with different names:

### Testing
Tests were executed with different setups, Central/Sensor logs analyzed, and results recorded in the tables below.

Tests were performed on an OCP cluster with 8 worker nodes, between each test the Central image metadata cache was cleared and image tables in central DB truncated.

For the first set of tests, 3 deployments with 30 replicas each were created, each deployment had different image names, but all the image names resolved to the same digest, for example:

```
ubi9:9.3
ubi9:9.3-1610
ubi9@sha256:66233eebd72bb5baa25190d4f55e1dc3fff3a9b77186c1f91a0abdb274452072
```

I waited before creating the 2nd and 3rd deployments until the first deployment was ready and scan caches populated.

Test # | Fully Qualified Images | Scanner | Feature | Fix | Unique Names | Sensor Sent Scan Requests | Central Scans 
:--:| :-: | :-: | :-: | :-:  | :-: | :-: | :-:
 1  | yes | off | off | 🚫 | 3 | 23 | 23
 2  | yes | on  | off | 🚫 | 3 | 2 | 2
 3  | yes | mix | off | 🚫 | 3 | 24 | 23
 4  | no  | off | on  | 🚫 | 6 | 33 | 31
 5  | no  | on  | on  | 🚫 | 6 | 24 | 1
 6  | no  | mix | on  | 🚫 | 6 | 25 | 22
 7  | no  | off | on  | ✅ | 6 | 6 | 4
 8  | no  | on  | on  | ✅ | 6 | 6 | 1
 9  | no  | mix | on  | ✅ | 6 | 6 | 2

For the next set of tests all 3 deployments were already running (from the steps above) - Central and Sensor were restarted at the same time, this put sensor into a state where it collected data waiting to re-connect to Central, after re-connecting recorded the results below:

Test # | Scanner | Feature | Fix | Unique Names | Sensor Sent Scan Requests | Central Scans 
:-: | :-: | :-: | :-: | :-: | :-: | :-: 
 10 | off | off | 🚫 | 3 | 68 | 69
 11 | on  | off | 🚫 | 3 | 1 | 5
 12 | off | on  | 🚫 | 6 | 58 | 59
 13 | on  | on  | 🚫 | 6 | 1 | 1
 14 | off | on  | ✅ | 6 | 3 | 2
 15 | on |  on  | ✅ | 6 | 2 | 2

 

Header | Desc
--- | ---
Scanner | `off` = scanner was scaled down to simulate errors, `on` = scanner was running, `mix` = scanner was off for the first deployment and on for the last two
Feature | `off` = the Unqualified Search Registries feature was disabled
Fix     | 🚫 = the main image was from `v4.4.3`, ✅ = the main image from this PR was running
Unique Names | # of unique names expected to see throughout the lifecycle of the tested deployments
Sensor Sent Scan Requests | # of requests sensor sent to central to scan an image
Central Scan | # of times Central attempted to scan an image

## Testing Takeaways / Observations

- The tests with fix tested successfully at limiting the number of scans sent to Central with the given test conditions.
   - These tests had the same (or less) number of requests to Central as unique image names (desired)
- At least one of the expected names was missing from the record in Central DB for all tests except for the ones with this PR's fix. 
   - This was caused by a data race in the Central's `ScanImageInternal` endpoint, there are no concurrency mechanisms protecting simultaneous scans for the same image, such that the last completed wins
   - This issue didn't affect tests with the fix enabled because Sensor was limiting requests to a single scan per image, the issue would still occur if multiple clusters were scanning the same image at the same time.
   - The fix only addresses one aspect of ACS's overall image scan caching, additional changes can (and should) be made to Central to increase likelihood of cache hits, reduce number of repeated scans, and eliminate data races.
- In larger environments there will be more variance due to more deployments / status updates / etc.
- Reminder: Scans will not be attempted on metadata errors, which are expected to occur before unqualified search registry feature changes the name of an image
- Test 13 was 'lucky' and Central Sent an updated image scan to Sensor before Sensor had a chance to send scan requests to Central, if re-ran would expect the sensor sent scan requests to be higher
- Raw data is available upon request

<details>
<summary>Additional Test Details</summary>
Logs were tailed with these conditions

```sh
oc logs -f -lapp=central | grep -viE "error enriching image|no signatures|this is here for now|registry_without_digest_call|updated existing image name|error fetching image signatures" --line-buffered | grep "ubi9"
oc logs -f -lapp=sensor | grep -iEv "Updated image name|not pullable" --line-buffered | grep -i "ubi9"
```

### Setup

Before each test in the initial set (1-9) did the following:

```sh
oc delete deploy ubia ubib ubic ubid
# clear central metadata cache with API call
# clear central image tables with SQL
# clear scanner layer tables with SQL (delete from layer; delete from imagetolayer)
oc delete pod -lapp=sensor
```

### Tests 1-3
```sh
# scanner on/off varied between tests
oc scale deploy/scanner --replicas=0
oc set env deploy/sensor ROX_UNQUALIFIED_SEARCH_REGISTRIES=false
oc set env deploy/central ROX_UNQUALIFIED_SEARCH_REGISTRIES=false
oc create deployment ubia --image=registry.access.redhat.com/ubi9:9.3-1610 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
# scanner on/off varied between tests, waited at this point until the deployment was cached
oc scale deploy/scanner --replicas=1
oc create deployment ubib --image=registry.access.redhat.com/ubi9:9.3 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
oc create deployment ubic --image=registry.access.redhat.com/ubi9@sha256:66233eebd72bb5baa25190d4f55e1dc3fff3a9b77186c1f91a0abdb274452072 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
```

### Tests 4-6

```sh
# scanner on/off varied between tests
oc set env deploy/sensor ROX_UNQUALIFIED_SEARCH_REGISTRIES=true
oc set env deploy/central ROX_UNQUALIFIED_SEARCH_REGISTRIES=true
oc create deployment ubia --image=ubi9:9.2-755.1697625012 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
# scanner on/off varied between tests, waited at this point until the deployment was cached
oc scale deploy/scanner --replicas=1
oc create deployment ubib --image=ubi9:9.2 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
oc create deployment ubic --image=ubi9@sha256:204383c3d96c0e6c7154c91d07764f92035738dd67aa8896679f7feb73f66bfd --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
```

### Tests 7-9
```sh
# scanner on/off varied between tests
oc set image deploy/central *=quay.io/dcaravel/sandbox/main:latest
oc set image deploy/sensor *=quay.io/dcaravel/sandbox/main:latest
oc set env deploy/sensor ROX_UNQUALIFIED_SEARCH_REGISTRIES=true
oc set env deploy/central ROX_UNQUALIFIED_SEARCH_REGISTRIES=true
oc create deployment ubia --image=ubi9:9.1.0-1817 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
# scanner on/off varied between tests, waited at this point until the deployment was cached
oc scale deploy/scanner --replicas=1
oc create deployment ubib --image=ubi9:9.1 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
oc create deployment ubic --image=ubi9@sha256:0bda22a725bb25c05b201ec250967a53fc4f3c6a7c82fcb2d8f6d3da3bed63dd --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
```

### Test 10-15

```sh
# scanner on/off varied between tests
oc scale deploy/scanner --replicas=0
# feature on/off varied between tests
oc set env deploy/sensor ROX_UNQUALIFIED_SEARCH_REGISTRIES=false
oc set env deploy/central ROX_UNQUALIFIED_SEARCH_REGISTRIES=false
oc create deployment ubia --image=ubi9:9.2-755.1697625012 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
oc create deployment ubib --image=ubi9:9.2 --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"
oc create deployment ubic --image=ubi9@sha256:204383c3d96c0e6c7154c91d07764f92035738dd67aa8896679f7feb73f66bfd --replicas=30 -- /bin/sh -c "trap : TERM INT; sleep infinity & wait"

oc delete pod -lapp=sensor && oc delete pod -lapp=central
```

### CODE REVIEW

File: sensor/common/detector/enricher.go [hunk 1]
```diff
@@ -143,6 +143,15 @@ outer:
 }
 
 func (c *cacheValue) scanAndSet(ctx context.Context, svc v1.ImageServiceClient, req *scanImageRequest) {
+	if features.UnqualifiedSearchRegistries.Enabled() {
+		// Guarding this with a feature flag because it's planned to be introduced in a
```
Comment (jvdm) [inline sensor/common/detector/enricher.go:147]: Nit: A JIRA to track this would be ideal.

File: sensor/common/detector/enricher.go [hunk 2]
```diff
@@ -170,6 +179,55 @@ func (c *cacheValue) scanAndSet(ctx context.Context, svc v1.ImageServiceClient,
 	c.image = scannedImage.GetImage()
 }
 
+func (c *cacheValue) scanAndSetWithLock(ctx context.Context, svc v1.ImageServiceClient, req *scanImageRequest) {
+	defer c.signal.Signal()
+
+	// A cacheValue is unique per image, obtain the lock before scanning to avoid sending multiple
+	// scan requests for the same image name at the same time.
+	c.lock.Lock()
+	defer c.lock.Unlock()
```
Comment (jvdm) [inline sensor/common/detector/enricher.go:188]: Do we add this `cacheValue` to the cache before scanning? I am thinking of the scenario of each scan request having its value until one of them is added to the cache, and then this lock would effectively function.

Comment (dcaravel) [reply sensor/common/detector/enricher.go:188]: Yes it is added to the cache before scanning with this line: 

https://github.com/stackrox/stackrox/blob/84014bb2026db2a7edeb6945f29eeaa9b25dee20/sensor/common/detector/enricher.go#L234

File: sensor/common/detector/enricher.go (hunk 1 above)
Comment (dcaravel) [reply sensor/common/detector/enricher.go]: https://github.com/stackrox/stackrox/pull/11359/commits/3c26bd345a045d153a6653d0eb294ece21ddfd13

**Review by BradLugo — COMMENTED** [review]
LGTM. Do we need to add something to the CHANGELOG? What other documentation would a fix like this need to be added to?

**Comment by dcaravel** [comment]
> LGTM. Do we need to add something to the CHANGELOG? What other documentation would a fix like this need to be added to?

Yes change log and release notes will be needed, I haven't updated the changelog yet (delaying to avoid dealing with frequent merge conflicts)

**Review by BradLugo — APPROVED** [review]

**Comment by dcaravel** [comment]
FYI, to include this PR in the 4.4.4 patch (when the time comes), the following is needed to avoid compilation errors.  This change has already been tested and the current unit tests will cover the code path.

Reason: the changes to `protoutils` [introduced here](https://github.com/stackrox/stackrox/pull/10156) are not in the 4.4. branch, therefore we have to 'fallback' to the previous util method when determining unique image names.  

```
diff --git a/sensor/common/detector/enricher.go b/sensor/common/detector/enricher.go
index c7a4195acc..ca87b59dbb 100644
--- a/sensor/common/detector/enricher.go
+++ b/sensor/common/detector/enricher.go
@@ -16,6 +16,7 @@ import (
        "github.com/stackrox/rox/pkg/expiringcache"
        "github.com/stackrox/rox/pkg/features"
        "github.com/stackrox/rox/pkg/images/types"
+       "github.com/stackrox/rox/pkg/images/utils"
        "github.com/stackrox/rox/pkg/protoutils"
        "github.com/stackrox/rox/pkg/set"
        "github.com/stackrox/rox/sensor/common/clusterid"
@@ -226,7 +227,7 @@ func (c *cacheValue) updateImageNoLock(image *storage.Image) {
                return
        }
 
-       c.image.Names = protoutils.SliceUnique(append(c.image.GetNames(), existingNames...))
+       c.image.Names = utils.UniqueImageNames(c.image.GetNames(), existingNames)
 }
```

**Review by RTann — APPROVED** [review]
I am a little concerned about the reduction in concurrency we'll now get, but I'm ok with this being merged behind a feature flag. We need to revisit how we do things because it's getting messy and confusing.

Also, are there any Jiras for the following?

```
This was caused by a data race in the Central's ScanImageInternal endpoint, there are no concurrency mechanisms protecting simultaneous scans for the same image, such that the last completed wins
This issue didn't affect tests with the fix enabled because Sensor was limiting requests to a single scan per image, the issue would still occur if multiple clusters were scanning the same image at the same time.
The fix only addresses one aspect of ACS's overall image scan caching, additional changes can (and should) be made to Central to increase likelihood of cache hits, reduce number of repeated scans, and eliminate data races.
```

**Comment by dcaravel** [comment]
> I am a little concerned about the reduction in concurrency we'll now get, but I'm ok with this being merged behind a feature flag. We need to revisit how we do things because it's getting messy and confusing.

I was concerned about this as well until I realized the `cacheValue` signal essentially does the same thing for the 'initial' scan (blocks until the request complete), this change extends that 'block' but for each unique name instead of just the 'first one'.

But just in case, putting it behind the feature flag is meant to be the 'off switch' should this cause unforeseen issues.

**Comment by dcaravel** [comment]
> Also, are there any Jiras for the following?

Not yet, will be soon, I'm working on a writeup summarizing the issues identified during this research, and with that will create the associated JIRAs.

**Pull request merged**